src/engine/renderer/glsl_source/computeLight_fp.glsl

```diff
@@ -76,15 +84,18 @@ void computeLight( vec3 lightDir, vec3 normal, vec3 viewDir, vec3 lightColor,
   // if no r_halfLambertLighting and no r_wrapAroundLighting.
   float NdotL = dot( normal, lightDir );
 
+#if !defined(USE_BSP_SURFACE)
 #if defined(r_halfLambertLighting)
   // http://developer.valvesoftware.com/wiki/Half_Lambert
   NdotL = NdotL * 0.5 + 0.5;
   NdotL *= NdotL;
-#elif defined(r_wrapAroundLighting)
+#endif
+#if defined(r_wrapAroundLighting)
```

Comment:
I think it's intentional that these are mutually exclusive. It doesn't make sense to use the two formulas manipulating normals together.

Maybe wrapAroundLighting should just be nuked? The cvar is marked cheat so it can't be used anyway. Also the code here seems wrong (using r_wrapAroundLighting as a number - probably it should be u_wrapAroundLighting)

Reply:
> The cvar is marked cheat so it can't be used anyway. 

The cvar is marked cheat so it can't be _disabled_.

Reply:
This is how it is set:

```c++
r_wrapAroundLighting = Cvar_Get( "r_wrapAroundLighting", "0.7", CVAR_CHEAT | CVAR_LATCH );
```

Reply:
In `tr_shader.cpp`, the `wrapAroundLighting` keyword is expected to be set on material stages.

Reply:
RBQUAKE3 (latest version of XreaL) use a variable indead (`u_LightWrapAround`):

https://github.com/RobertBeckebans/RBQUAKE-3/blob/9559591679085e5462de80049990a6a302616b94/base/glsl/forwardLighting_fp.glsl#L1037-L1041

Reply:
We do that, though:

```
	if ( r_wrapAroundLighting->value )
		AddConst( str, "r_wrapAroundLighting", r_wrapAroundLighting->value );
```

RBQUAKE3 does this like us:

```
r_wrapAroundLighting  = ri.Cvar_Get( "r_wrapAroundLighting", "0.7", CVAR_CHEAT | CVAR_SHADER );
```

I assume the value is set in a similar way thanks to the `CVAR_SHADER` flag, but then no code makes use of this value in RBQUAKE3.

Reply:
I guess the cvar being a float is a relic of a time it was being prototyped, the value is expected to be set by materials.

Reply:
Our legacy `forwardLighting_fp.glsl` also uses `u_LightWrapAround`:

```c++
float NL = clamp(dot(normal, lightDir) + u_LightWrapAround, 0.0, 1.0) / clamp(1.0 + u_LightWrapAround, 0.0, 1.0);
```

Reply:
So, I modified the code to use the variable, so the feature is only used on materials setting it, but I disabled the cvar by default. I removed the cheat flag as I assume it was part of a prototype, like the float value (that is now expected to be `0`/`1` to disable/enable).

Reply:
OK but it should still be mutually exclusive with half lambert

Reply:
I still don't get why this couldn't be cumulative.

Also doing either one or the other is not decidable: we have no way to know in this code if a material uses wrapAround or not.

Reply:
Well if you really wanted to do both at once, the only way that would make sense is to do "wrap around" first.



> Also doing either one or the other is not decidable: we have no way to know in this code if a material uses wrapAround or not.

Well the old way was just do disable wraparound if halflambert is enabled.

Reply:
> The old way was just to disable wraparound if halflambert is enabled.

That is not per-material.

Things like halfLambert and wrapAround are to meant to be used on some surfaces of the game in the same scene.

Reply:
I might suggest once again nuking wraparound as it causes us problems for no benefit.

I don't see any way to set half-lambert per material. It seems to be a global cvar only. So that's how it is possible to disable wraparound whenever half-lambert is on.

By the way if half-lambert is disabled for BSP surfaces, then wraparound probably should be too for the same reasons.

Reply:
I sent an e-mail to Tr3b to know what wrapAround was meant to be in XreaL to begin with.

I noticed something is wrong in that code anyway, as I believe `clamp(1.0 + u_LightWrapAround, 0.0, 1.0)` is always `1`.

Reply:
In [that page](https://www.jordanstevenstechart.com/lighting-models) the writer also talk about a variant of Half-Lambert that is named `Diffuse Wrap`, which allows to use an arbitrary value instead of 0.5:

```glsl
pow(NdotL * wrapValue + (1-wrapValue),2) 
```

We may allow to customize the `wrapValue` in materials like wrapAround does.

Reply:
> I noticed something is wrong in that code anyway, as I believe `clamp(1.0 + u_LightWrapAround, 0.0, 1.0)` is always `1`.

Not if it's negative.

Reply:
wrapAroundLighting has nuked:

- https://github.com/DaemonEngine/Daemon/pull/1015